spark-3.1-spanner-lib/src/main/java/com/google/cloud/spark/spanner/BatchClientWithCloser.java

```diff
@@ -27,8 +26,17 @@ public BatchClientWithCloser(Spanner spanner, BatchClient batchClient) {
     this.batchClient = batchClient;
   }
 
+  /*
+   * close is a runtime hook for AutoCloseable to properly shut down resources
+   * before this object is garbage collected. It is useful in scenarios such as
+   * asynchronous processing for which we won't have a deterministic time/scope
+   * for when a Spanner object will be closed.
+   */
   @Override
   public void close() {
```

Comment:
QQ: Is the close() the last function called in the class ? I just don't understand how the close() is called multiple times.

Reply:
1. It is called indeed last by Spark
2. If the code isn't properly closed, the AutoCloseable becomes useful

Reply:
3. @halio-g multiple readers are created by Spark hence why you'd see many missing Spanner.shutdown calls